### PR TITLE
Do not ban peers for sending multiple valid requests

### PR DESCRIPTION
**Pull request by bkchr**
We introduced banning of peers who spam us with the same request (more
than 2 times). However, we missed that it is completely legal to send
the same request multiple times as long as we did not provide any
answer. An example for that is the justification request. This request
is send multiple times until we could fetch the justification from one
of our peers. So, the solution to this problem is to tag requests as
fulfilled and to start counting these fulfilled requests. If the number
is higher than what we allow, the peer should be banned.

Thank you for your Pull Request!

### CODE REVIEW

File: client/network/src/block_request_handler.rs [hunk 1]
```diff
@@ -212,6 +228,21 @@ impl<B: BlockT> BlockRequestHandler<B> {
 				max_blocks,
 			)?;
 
+			// If any of the blocks contains nay data, we can consider it as successful request.
```
Comment (arkpar) [inline client/network/src/block_request_handler.rs:231]:
```suggestion
			// If any of the blocks contains any data, we can consider it as successful request.
```

**Comment by arkpar** [comment]
> We introduced banning of peers who spam us with the same request (more
than 2 times). However, we missed that it is completely legal to send
the same request multiple times as long as we did not provide any
answer.

You mean it is legal to send the same request multiple times when we provide an empty answer?

The sync protocol requires that each request will either be answered (possible with empty block list), or the connection to be terminated.

**Comment by bkchr** [comment]
> You mean it is legal to send the same request multiple times when we provide an empty answer?

Yes. This for example happens for the justification request.

**Review by arkpar — APPROVED** [review]

**Review by andresilva — APPROVED** [review]

**Comment by andresilva** [comment]
bot merge force

**Pull request merged**